### PR TITLE
Add llvm support for s390x, fix xla and aot issue

### PR DESCRIPTION
**Pull request by Sidong-Wei**
Fixes issue #39912 . 

Currently, 'systemz' or 's390x' was not included in the llvm build target list. It will throw an warning saying "'z14' is not a recognized processor for this target (ignoring processor)" then fail the endianness-check when using xla compiling.

This patch adds llvm support for s390x machine, fixing xla (including aot compile) features. It fixes 8 failed test cases and 1 failed-to-build test case on s390x. It is also has been tested to have no effect on x86 machines.

### CODE REVIEW

**Comment by Sidong-Wei** [comment]
Hi @joker-eph , would you mind taking a look into my PR? It has been awaiting review for a while, thank you.

**Review by joker-eph — DISMISSED** [review]

**Comment by joker-eph** [comment]
Sorry for the delay: I am auto-subscribed to PR touching tensorflow/compiler but I am mostly looking into MLIR-related work

**Comment by Sidong-Wei** [comment]
Thanks a lot for the approval, appreciate it.

**Comment by joker-eph** [comment]
@sanjoy who's the best point of contact for tfcompile here?

**Comment by Sidong-Wei** [comment]
There seems to be a new conflict here (removing alias), I will double-check and push again soon.

File: tensorflow/compiler/xla/service/cpu/BUILD
Comment (Sidong-Wei) [reply tensorflow/compiler/xla/service/cpu/BUILD:194]: I'm not sure, I only verify that it could successfully build on s390x machine (and has no effect on x86_64), JIT compile could work and test cases including `//tensorflow/python/compiler/xla:jit_test` will pass. But there still a number of failures in `//tensorflow/compiler/` unit test module. Do you need me to do any other verification?

File: third_party/llvm/llvm.autogenerated.BUILD
Comment (Sidong-Wei) [reply third_party/llvm/llvm.autogenerated.BUILD:537]: Sounds good, I was guessing maybe there's a script to generate these lines. Thank you.

**Comment by gbaned** [comment]
@sanjoy Any update on this PR? Please. Thanks!

**Comment by sanjoy** [comment]
> @sanjoy Any update on this PR? Please. Thanks!

The PR LGTM, but I want to not merge it as-is (since I need to make some internal changes).  How do I do that?  If I approve it, it will get auto-merged right?

**Review by rthadur — DISMISSED** [review]

**Review by sanjoy — CHANGES_REQUESTED** [review]
Hi,

https://github.com/tensorflow/tensorflow/commit/c9d13afacd601789d0c71152f257c1b6a7a599e8 adds the SystemZ target to `llvm.autogenerated.BUILD`.  But when importing this CL, I noticed another issue: the dependency on `"@llvm-project//llvm:SystemZCodeGen"` cannot be unconditional, it needs to be guarded by a macro like `if_llvm_aarch64_available`.  Once you fix that this CL should be good to merge.

The implementation of the `if_llvm_aarch64_available ` macro is unfortunately [stubbed out](https://github.com/tensorflow/tensorflow/blob/1186e3f2098793952aa82bf356dfe51b967fb26c/tensorflow/core/platform/default/build_config.bzl#L778), but you can use `tf_windows_aware_platform_deps` as an example of how to use `select` to conditionally enable the dependency.

**Comment by Sidong-Wei** [comment]
Hi @sanjoy ,
thanks for the comment. I notice that `@llvm-project//llvm:SystemZCodeGen` is already guarded in `tensorflow/compiler/xla/service/cpu/BUILD`, so I add a similar guard to it in `tensorflow/compiler/aot/BUILD`(They seem to be the only 2 places that this dependency is used). I use the `//tensorflow:linux_s390x` macro to guard for s390x machine, and I leave the `if_llvm_aarch64_available` part "as is" because you mention it needs to be fixed in the future. Please let me know if I missed anything or if you prefer I use another macro, thanks.

File: third_party/llvm/llvm.autogenerated.BUILD
Comment (sanjoy) [reply third_party/llvm/llvm.autogenerated.BUILD:537]: This bit should be gone after you rebase on head.

File: tensorflow/compiler/aot/compile.cc
Comment (sanjoy) [inline tensorflow/compiler/aot/compile.cc:182]: This needs to be `#ifdef` guarded, like `TF_LLVM_AARCH64_AVAILABLE`.

File: tensorflow/compiler/aot/BUILD
Comment (sanjoy) [inline tensorflow/compiler/aot/BUILD:80]: This needs to be in a macro, like `if_llvm_aarch64_available`.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/aot/BUILD
Comment (Sidong-Wei) [reply tensorflow/compiler/aot/BUILD:80]: ok, I will put the "select" part in a macro under `tensorflow/core/platform/default/build_config.bzl` then use it here.

File: tensorflow/compiler/aot/compile.cc
Comment (Sidong-Wei) [reply tensorflow/compiler/aot/compile.cc]: This should be solved now.

**Comment by Sidong-Wei** [comment]
Thank you @sanjoy for the feedback, I have wrote a macro `if_llvm_s390x_available` and use it exactly the same as `if_llvm_aarch64_available`. Please let me know if there's anything wrong with it.

**Review by sanjoy — DISMISSED** [review]
Thanks, this should work.

**Comment by Sidong-Wei** [comment]
Thank you for the approval @sanjoy , I'm so sorry that I made a typo in my last commit, I think that could be the reason causing the CI tests failed. I have fixed it now, could you please re-approve it to see if the CI could pass this time?

**Review by sanjoy — DISMISSED** [review]

**Comment by Sidong-Wei** [comment]
Hmm, it seems that it is still failing, I will look into this tomorrow, thank you @sanjoy .

**Comment by Sidong-Wei** [comment]
I think the duplicate dependency was the cause of the failure, it was actually a collision but was not detected by git. It is fixed now and should be able to pass the CI test (I successfully built it on my side).

**Review by sanjoy — DISMISSED** [review]

**Review by rthadur — DISMISSED** [review]

**Comment by Sidong-Wei** [comment]
CI build has passed but sanity check failed. It suggested me to do the following change:
```
diff --git a/tensorflow/compiler/aot/BUILD b/tensorflow/compiler/aot/BUILD
index 33d0adc96d..33edb527f4 100644
--- a/tensorflow/compiler/aot/BUILD
+++ b/tensorflow/compiler/aot/BUILD
@@ -36,8 +36,9 @@ cc_library(
         "flags.h",
         "quantize.h",
     ],
-    defines = if_llvm_aarch64_available(["TF_LLVM_AARCH64_AVAILABLE=1"])
-              + if_llvm_s390x_available(["TF_LLVM_S390X_AVAILABLE=1"]),
+    defines = if_llvm_aarch64_available(["TF_LLVM_AARCH64_AVAILABLE=1"]) + if_llvm_s390x_available([
+        "TF_LLVM_S390X_AVAILABLE=1",
+    ]),
     visibility = ["//tensorflow/python:__pkg__"],
     deps = [
         ":aot_only_var_handle_op",
```

However, this does not really look neat to me, and I am not familiar with the preferred style here either. Do I need to make this change or if you could suggest a better style to pass sanity check here? Thanks.

**Comment by sanjoy** [comment]
> However, this does not really look neat to me, and I am not familiar with the preferred style here either. Do I need to make this change or if you could suggest a better style to pass sanity check here?

I would suggest making the change, I don't think fighting the build-file formatter is a good use of our time.

**Comment by gbaned** [comment]
@Sidong-Wei Can you please check @sanjoy's comments and address Ubuntu Sanity errors? Thanks!

**Comment by Sidong-Wei** [comment]
Sure @gbaned , I have pushed the change according to the sanity check suggestion. I just wanted to check if there's any better idea than its suggestion, it should be good now.

**Review by sanjoy — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Comment by Sidong-Wei** [comment]
Hi, it seems to me that all checks have passed, just wondering is there anything else I need to do before the PR is merged?

**Pull request merged**
